pkg/sql/opt_catalog.go

```diff
@@ -611,7 +611,7 @@ func newOptTable(
 	// Create the table's column mapping from sqlbase.ColumnID to column ordinal.
 	ot.colMap = make(map[sqlbase.ColumnID]int, ot.ColumnCount())
 	for i, n := 0, ot.ColumnCount(); i < n; i++ {
-		ot.colMap[sqlbase.ColumnID(ot.Column(i).ColID())] = i
+		ot.colMap[sqlbase.ColumnID(ot.Column(i).PhysicalColID())] = i
```

Comment:
What error happens if this isn't here? I'm very dubious about this line.

Reply:
This logic test fails for ALTER COLUMN TYPE with `requested column y not in index`

Alter column type creates a new column with the new type and "swaps" it with the original column. There's a state where we set the LogicalColumnID of the new column to the LogicalColumnID of the original column (which in this test case happens to be the physical ColumnID as well). However the original column is dropped and there is no more column with the column of the original physical ColumnID, I think this mapping maps columns to their physical column id so when it looks up using the LogicalColumnID (which is the original columns physical ColumnID) there is no column. Correct me if I'm wrong or if this explanation is too confusing.

From what I understand, colMap should actually map the physical ColumnID of the column to the ordinal.

```	
// colMap is a mapping from unique ColumnID to column ordinal within the
	
// table. This is a common lookup that needs to be fast.
	
colMap map[sqlbase.ColumnID]int
``` 

```
statement ok
ALTER TABLE t13 ALTER COLUMN y TYPE STRING

query IT
SELECT x, y FROM t13 ORDER BY x
----
1  1
2  2
```

Reply:
@RaduBerinde can you comment on this? I don't know enough.